libsolidity/formal/SMTChecker.cpp

```diff
@@ -1901,7 +1904,16 @@ FunctionDefinition const* SMTChecker::inlinedFunctionCallToDefinition(FunctionCa
 		return nullptr;
 
 	FunctionType const& funType = dynamic_cast<FunctionType const&>(*_funCall.expression().annotation().type);
-	if (funType.kind() != FunctionType::Kind::Internal)
+	if (funType.kind() == FunctionType::Kind::External)
+	{
+		auto memberAccess = dynamic_cast<MemberAccess const*>(&_funCall.expression());
+		auto identifier = memberAccess ?
+			dynamic_cast<Identifier const*>(&memberAccess->expression()) :
+			nullptr;
+		if (!identifier || identifier->name() != "this")
```

Comment:
As we learned, the following code is legal:
```
SomeOtherContract this = bla();
this.f();
```
So I think you also have to check the type of `identifier->referencedDeclaration()`.

Reply:
Oh, true...
Even if the type is `ContractDefintion`, can we ever be sure it's the `this` we think it is?
```
contract C
{
    uint public x;
    C c;
    function f(C _c) public {
        c = _c;
    }
    function g() public {
        C this = c;
        this.h();
    }
    function h() public {
        x = 2;
    }
}
```

Reply:
I think checking for MagicVariableDeclaration should be enough.